File: ai-research-copilot/backend/app/schemas/workflow.py
"""
Workflow Pydantic v2 schemas.

Request/response schemas for workflow CRUD, execution, and monitoring.
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class WorkflowCreate(BaseModel):
    """Schema for creating a new workflow."""

    model_config = ConfigDict(str_strip_whitespace=True)

    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Human-readable workflow name.",
    )
    description: Optional[str] = Field(
        default=None,
        max_length=2048,
        description="Optional description of the workflow purpose and behavior.",
    )
    workflow_type: str = Field(
        ...,
        max_length=30,
        description="Workflow execution pattern (sequential, parallel, conditional, loop).",
    )
    definition: dict = Field(
        ...,
        description="Workflow graph structure with nodes and edges.",
    )
    is_template: bool = Field(
        default=False,
        description="Whether this workflow should be saved as a reusable template.",
    )


class WorkflowUpdate(BaseModel):
    """Schema for updating an existing workflow."""

    model_config = ConfigDict(str_strip_whitespace=True)

    name: Optional[str] = Field(
        default=None,
        min_length=1,
        max_length=255,
        description="Updated workflow name.",
    )
    description: Optional[str] = Field(
        default=None,
        max_length=2048,
        description="Updated workflow description.",
    )
    workflow_type: Optional[str] = Field(
        default=None,
        max_length=30,
        description="Updated workflow execution pattern.",
    )
    definition: Optional[dict] = Field(
        default=None,
        description="Updated workflow graph structure.",
    )
    status: Optional[str] = Field(
        default=None,
        max_length=30,
        description="Updated workflow status (draft, active, archived).",
    )
    is_template: Optional[bool] = Field(
        default=None,
        description="Updated template flag.",
    )


class WorkflowResponse(BaseModel):
    """Schema for workflow data returned in API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    name: str
    workflow_type: str
    status: str
    is_template: bool
    execution_count: int
    created_at: datetime
    updated_at: datetime


class WorkflowDetail(WorkflowResponse):
    """Extended workflow schema with full definition and metadata."""

    description: Optional[str] = None
    definition: dict
    last_executed_at: Optional[datetime] = None
    metadata_extra: Optional[dict] = None


class WorkflowList(BaseModel):
    """Paginated list of workflows."""

    model_config = ConfigDict(from_attributes=True)

    items: list[WorkflowResponse]
    total: int
    page: int
    page_size: int


class WorkflowExecuteRequest(BaseModel):
    """Schema for executing a workflow."""

    model_config = ConfigDict(str_strip_whitespace=True)

    input_data: Optional[dict] = Field(
        default=None,
        description="Input data to pass to the workflow execution.",
    )


class WorkflowExecutionResponse(BaseModel):
    """Schema for workflow execution data returned in API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workflow_id: UUID
    user_id: UUID
    status: str
    input_data: Optional[dict] = None
    output_data: Optional[dict] = None
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    current_step: Optional[str] = None
    progress: float
    created_at: datetime


class WorkflowExecutionList(BaseModel):
    """Paginated list of workflow executions."""

    model_config = ConfigDict(from_attributes=True)

    items: list[WorkflowExecutionResponse]
    total: int


class WorkflowStepExecutionResponse(BaseModel):
    """Schema for individual workflow step execution data."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    execution_id: UUID
    step_name: str
    step_type: str
    status: str
    input_data: Optional[dict] = None
    output_data: Optional[dict] = None
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    retry_count: int

File: ai-research-copilot/frontend/src/types/workflow.ts
export interface Workflow {
  id: string;
  name: string;
  description: string | null;
  status: "draft" | "active" | "paused" | "archived";
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  trigger_type: string | null;
  execution_count: number;
  last_executed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface WorkflowDetail extends Workflow {
  executions: WorkflowExecution[];
}

export interface WorkflowNode {
  id: string;
  type: string;
  label: string;
  config: Record<string, unknown>;
  position: { x: number; y: number };
}

export interface WorkflowEdge {
  id: string;
  source: string;
  target: string;
  label?: string;
}

export interface WorkflowExecution {
  id: string;
  workflow_id: string;
  status: "pending" | "running" | "completed" | "failed" | "cancelled";
  input_data: Record<string, unknown> | null;
  output_data: Record<string, unknown> | null;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
  created_at: string;
}

export interface WorkflowExecuteRequest {
  input_data: Record<string, unknown>;
}

export interface WorkflowState {
  workflows: Workflow[];
  currentWorkflow: WorkflowDetail | null;
  isLoading: boolean;
  error: string | null;
  setWorkflows: (workflows: Workflow[]) => void;
  setCurrentWorkflow: (workflow: WorkflowDetail | null) => void;
  setIsLoading: (isLoading: boolean) => void;
  setError: (error: string | null) => void;
}
